dtm0/fop.c

```diff
@@ -0,0 +1,392 @@
+#include "lib/trace.h"
+#include "addb2/addb2.h"
+#include "be/dtm0_log.h"
+#include "dtm0/addb2.h"
+#include "dtm0/fop.h"
+#include "dtm0/fop_xc.h"
+#include "dtm0/service.h"
+#include "dtm0/tx_desc.h"
+#include "dtm0/tx_desc_xc.h"
+#include "lib/errno.h"
+#include "lib/assert.h"
+#include "lib/memory.h"
+#include "lib/chan.h"
+#include "lib/finject.h"
+#include "lib/time.h"
+#include "lib/misc.h"           /* M0_IN() */
+#include "fop/fop.h"
+#include "fop/fom.h"
+#include "fop/fom_generic.h"
+#include "fop/fop_item_type.h"
+#include "rpc/rpc_opcodes.h"
+#include "rpc/rpc.h"
+#include "rpc/rpclib.h"
+
+static void dtm0_rpc_item_reply_cb(struct m0_rpc_item *item);
+
+/*
+  RPC item operations structures.
+ */
+const struct m0_rpc_item_ops dtm0_req_fop_rpc_item_ops = {
+        .rio_replied = dtm0_rpc_item_reply_cb,
+};
+
+struct m0_fop_type dtm0_req_fop_fopt;
+struct m0_fop_type dtm0_rep_fop_fopt;
+
+/*
+  Fom specific routines for corresponding fops.
+ */
+static int dtm0_fom_tick(struct m0_fom *fom);
+static int dtm0_fom_create(struct m0_fop *fop, struct m0_fom **out,
+			       struct m0_reqh *reqh);
+static void dtm0_fom_fini(struct m0_fom *fom);
+static size_t dtm0_fom_locality(const struct m0_fom *fom);
+
+static const struct m0_fom_ops dtm0_req_fom_ops = {
+	.fo_fini = dtm0_fom_fini,
+	.fo_tick = dtm0_fom_tick,
+	.fo_home_locality = dtm0_fom_locality
+};
+
+extern struct m0_reqh_service_type dtm0_service_type;
+
+enum dtm0_phases {
+	DTM0_REQ_WHATEVER = M0_FOPH_NR + 1,
+};
+
+static const struct m0_fom_type_ops dtm0_req_fom_type_ops = {
+        .fto_create = dtm0_fom_create,
+};
+
+static void dtm0_rpc_item_reply_cb(struct m0_rpc_item *item)
+{
+	struct m0_fop *reply = NULL;
+
+	M0_ENTRY("item=%p", item);
+
+        M0_PRE(item != NULL);
+	M0_PRE(M0_IN(m0_fop_opcode(m0_rpc_item_to_fop(item)),
+		     (M0_DTM0_REQ_OPCODE)));
+
+	if (m0_rpc_item_error(item) == 0) {
+		reply = m0_rpc_item_to_fop(item->ri_reply);
+		M0_ASSERT(M0_IN(m0_fop_opcode(reply), (M0_DTM0_REP_OPCODE)));
+	}
+
+	M0_LEAVE("reply=%p", reply);
+}
+
+M0_INTERNAL void m0_dtm0_fop_fini(void)
+{
+	m0_fop_type_fini(&dtm0_req_fop_fopt);
+	m0_fop_type_fini(&dtm0_rep_fop_fopt);
+	m0_xc_dtm0_fop_fini();
+}
+
+enum {
+	M0_FOPH_DTM0_LOGGING = M0_FOPH_TYPE_SPECIFIC,
+};
+
+struct m0_sm_state_descr dtm0_phases[] = {
+	[M0_FOPH_DTM0_LOGGING] = {
+		.sd_name      = "logging",
+		.sd_allowed   = M0_BITS(M0_FOPH_SUCCESS,
+					M0_FOPH_FAILURE)
+	},
+};
+
+struct m0_sm_trans_descr dtm0_phases_trans[] = {
+	[ARRAY_SIZE(m0_generic_phases_trans)] =
+	{"dtm0_1-fail", M0_FOPH_TYPE_SPECIFIC, M0_FOPH_FAILURE},
+	{"dtm0_1-success", M0_FOPH_TYPE_SPECIFIC, M0_FOPH_SUCCESS},
+};
+
+static struct m0_sm_conf dtm0_conf = {
+	.scf_name      = "dtm0-fom",
+	.scf_nr_states = ARRAY_SIZE(dtm0_phases),
+	.scf_state     = dtm0_phases,
+	.scf_trans_nr  = ARRAY_SIZE(dtm0_phases_trans),
+	.scf_trans     = dtm0_phases_trans,
+};
+
+M0_INTERNAL int m0_dtm0_fop_init(void)
+{
+	static int init_once = 0;
```

Comment:
A couple of things like this can also be found in other parts of the MERO but strictly speaking, some kind of atomicity or compiler barrier would be good here.

Reply:
@sergey-shilov , please put a comment here, so that it can be found in future and fix.
Proposal: "DTM0_TECH_DEBT: ... "

Reply:
Done.